### 📥 Leitura dos Dados

In [2]:
import pandas as pd
import numpy as plt

df = pd.read_csv('./aluguel_imoveis.csv')

df.head()

,bairro,tipo_imovel,mobiliado,elevador,aceita_pet,metragem_m2,quartos,banheiros,vagas,idade_imovel,andar,dist_metro_km,condominio,valor_aluguel
0,Zona Sul,Studio,Sim,Sim,Sim,43.8,3,2,0,31,16,3.12,359.01,4891.45
1,Zona Oeste,Apartamento,Sim,Sim,Sim,81.8,1,2,1,44,11,NaN,367.50,5352.06
2,Zona Leste,Apartamento,Não,Sim,Não,77.3,2,1,1,36,1,3.20,960.67,3236.37
3,Zona Norte,Apartamento,Sim,Sim,Sim,93.9,2,2,2,35,5,3.79,270.02,4697.57
4,Zona Sul,Apartamento,Não,Sim,Não,50.2,2,1,1,20,8,1.42,918.23,5669.06


In [3]:
df.describe()

# Dados inconsistentes e Negativos

,metragem_m2,quartos,banheiros,vagas,idade_imovel,andar,dist_metro_km,condominio,valor_aluguel
count,1140.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1140.000000,1140.000000,1164.000000
mean,83.184825,2.182500,1.637500,1.066667,22.015000,12.226667,2.066237,546.219193,5163.538900
std,357.705203,0.831527,0.640132,0.794406,12.813597,9.116535,1.341700,248.185622,1902.235469
min,-104.100000,1.000000,1.000000,0.000000,-13.000000,0.000000,-4.120000,2.350000,-2000.000000
25%,52.200000,2.000000,1.000000,1.000000,12.000000,6.000000,1.020000,376.035000,3869.765000
50%,68.700000,2.000000,2.000000,1.000000,22.000000,12.000000,1.925000,545.330000,4839.215000
75%,85.700000,3.000000,2.000000,2.000000,33.000000,18.000000,2.922500,724.272500,6093.030000
max,9130.000000,4.000000,3.000000,3.000000,44.000000,120.000000,6.280000,1325.850000,15806.070000


In [4]:
df_clean = df

df_clean = df_clean[(df_clean['metragem_m2'] > 1) &
          (df_clean['idade_imovel'] > 1) & 
          (df_clean['valor_aluguel'] > 1) &
          (df_clean['dist_metro_km'] > 0)]


df_clean.describe()

,metragem_m2,quartos,banheiros,vagas,idade_imovel,andar,dist_metro_km,condominio,valor_aluguel
count,990.000000,990.000000,990.000000,990.000000,990.000000,990.000000,990.000000,941.000000,990.000000
mean,85.924444,2.166667,1.643434,1.066667,23.439394,12.184848,2.096545,549.731233,5204.675990
std,383.547864,0.834966,0.638448,0.802927,12.017760,9.441490,1.326583,247.020464,1879.875887
min,1.500000,1.000000,1.000000,0.000000,2.000000,0.000000,0.010000,6.430000,1689.980000
25%,51.825000,2.000000,1.000000,1.000000,14.000000,5.250000,1.025000,389.150000,3901.187500
50%,68.600000,2.000000,2.000000,1.000000,23.500000,12.000000,1.960000,545.030000,4871.785000
75%,85.775000,3.000000,2.000000,2.000000,33.000000,18.000000,2.977500,716.520000,6127.630000
max,9130.000000,4.000000,3.000000,3.000000,44.000000,120.000000,6.280000,1325.850000,15806.070000


## Tratando Variaveis categoricas

In [5]:
df_clean.info()

<class 'pandas.DataFrame'>
Index: 990 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   bairro         935 non-null    str    
 1   tipo_imovel    942 non-null    str    
 2   mobiliado      937 non-null    str    
 3   elevador       990 non-null    str    
 4   aceita_pet     990 non-null    str    
 5   metragem_m2    990 non-null    float64
 6   quartos        990 non-null    int64  
 7   banheiros      990 non-null    int64  
 8   vagas          990 non-null    int64  
 9   idade_imovel   990 non-null    int64  
 10  andar          990 non-null    int64  
 11  dist_metro_km  990 non-null    float64
 12  condominio     941 non-null    float64
 13  valor_aluguel  990 non-null    float64
dtypes: float64(4), int64(5), str(5)
memory usage: 116.0 KB


In [6]:

col_categoricas = df_clean.select_dtypes(include=['object', 'category']).columns.tolist()

col_numericas = df_clean.select_dtypes(include=['number']).columns.tolist()

print(f"Categóricas: {col_categoricas}")
print(f"Numéricas: {col_numericas}")

Categóricas: ['bairro', 'tipo_imovel', 'mobiliado', 'elevador', 'aceita_pet']
Numéricas: ['metragem_m2', 'quartos', 'banheiros', 'vagas', 'idade_imovel', 'andar', 'dist_metro_km', 'condominio', 'valor_aluguel']


C:\Users\joaoh\AppData\Local\Temp\ipykernel_49492\2832965561.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  col_categoricas = df_clean.select_dtypes(include=['object', 'category']).columns.tolist()


In [7]:
from sklearn.preprocessing import OneHotEncoder

enc = OneHotEncoder(sparse_output=False, drop=None)

colunas_transformadas = enc.fit_transform(df_clean[col_categoricas])

df_dummies = pd.DataFrame(
    colunas_transformadas, 
    columns=enc.get_feature_names_out(col_categoricas),
    index=df_clean.index # Mantém o índice original
)

df_final = pd.concat([df_clean[col_numericas], df_dummies], axis=1)

df_final.head()

,metragem_m2,quartos,banheiros,vagas,idade_imovel,andar,dist_metro_km,condominio,valor_aluguel,bairro_Centro,...,tipo_imovel_Cobertura,tipo_imovel_Studio,tipo_imovel_nan,mobiliado_Não,mobiliado_Sim,mobiliado_nan,elevador_Não,elevador_Sim,aceita_pet_Não,aceita_pet_Sim
0,43.8,3,2,0,31,16,3.12,359.01,4891.45,0.0,...,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0
2,77.3,2,1,1,36,1,3.20,960.67,3236.37,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0
3,93.9,2,2,2,35,5,3.79,270.02,4697.57,0.0,...,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0
4,50.2,2,1,1,20,8,1.42,918.23,5669.06,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0
5,70.9,3,1,1,11,4,5.97,467.32,5555.51,0.0,...,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0


In [8]:
cols_nan = [c for c in df_final.columns if 'nan' in c.lower()]
print(f"Colunas encontradas: {cols_nan}")

# Agora sim, apague usando a lista que o próprio Python encontrou
df_final = df_final.drop(columns=cols_nan)

df_final = df_final.dropna()

Colunas encontradas: ['bairro_nan', 'tipo_imovel_nan', 'mobiliado_nan']


In [9]:
pd.set_option('display.max_columns', None)

# Configura para mostrar todas as linhas
pd.set_option('display.max_rows', None)

# Agora chame o seu dataframe
df_final.head()

,metragem_m2,quartos,banheiros,vagas,idade_imovel,andar,dist_metro_km,condominio,valor_aluguel,bairro_Centro,bairro_Zona Leste,bairro_Zona Norte,bairro_Zona Oeste,bairro_Zona Sul,tipo_imovel_Apartamento,tipo_imovel_Cobertura,tipo_imovel_Studio,mobiliado_Não,mobiliado_Sim,elevador_Não,elevador_Sim,aceita_pet_Não,aceita_pet_Sim
0,43.8,3,2,0,31,16,3.12,359.01,4891.45,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0
2,77.3,2,1,1,36,1,3.20,960.67,3236.37,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
3,93.9,2,2,2,35,5,3.79,270.02,4697.57,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0
4,50.2,2,1,1,20,8,1.42,918.23,5669.06,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
5,70.9,3,1,1,11,4,5.97,467.32,5555.51,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0


In [10]:
df_final.info()

<class 'pandas.DataFrame'>
Index: 941 entries, 0 to 1199
Data columns (total 23 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   metragem_m2              941 non-null    float64
 1   quartos                  941 non-null    int64  
 2   banheiros                941 non-null    int64  
 3   vagas                    941 non-null    int64  
 4   idade_imovel             941 non-null    int64  
 5   andar                    941 non-null    int64  
 6   dist_metro_km            941 non-null    float64
 7   condominio               941 non-null    float64
 8   valor_aluguel            941 non-null    float64
 9   bairro_Centro            941 non-null    float64
 10  bairro_Zona Leste        941 non-null    float64
 11  bairro_Zona Norte        941 non-null    float64
 12  bairro_Zona Oeste        941 non-null    float64
 13  bairro_Zona Sul          941 non-null    float64
 14  tipo_imovel_Apartamento  941 non-null    

In [11]:
X = df_final.drop(columns=['valor_aluguel'])
y = df_final['valor_aluguel']

In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

modelo = LinearRegression()
modelo.fit(X_train, y_train)

# Fazendo previsões
y_pred = modelo.predict(X_test)

# Avaliando o modelo
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Erro Quadrático Médio (MSE): {mse:.2f}")
print(f"Coeficiente de Determinação (R²): {r2:.2f}")

Erro Quadrático Médio (MSE): 1216136.18
Coeficiente de Determinação (R²): 0.61


In [13]:
import statsmodels.api as sm
import pandas as pd

# X e y já limpos e com dummies
X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)

X_train = X_train.astype(float)
X_train = sm.add_constant(X_train)

model_sm = sm.OLS(y_train, X_train).fit()
print(model_sm.summary())

                            OLS Regression Results                            
Dep. Variable:          valor_aluguel   R-squared:                       0.687
Model:                            OLS   Adj. R-squared:                  0.678
Method:                 Least Squares   F-statistic:                     74.99
Date:                Mon, 23 Mar 2026   Prob (F-statistic):          2.65e-157
Time:                        11:02:49   Log-Likelihood:                -5920.7
No. Observations:                 705   AIC:                         1.188e+04
Df Residuals:                     684   BIC:                         1.198e+04
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                    2

Exercício 1 – Regressão Linear | Valor do Aluguel

Foi desenvolvido um modelo de regressão linear para prever o valor do aluguel a partir de características do imóvel. Os dados passaram por limpeza de valores ausentes, inconsistências e outliers, além da criação de dummies para variáveis categóricas. Utilizou-se separação holdout para treino e teste. O modelo final foi escolhido com base em métricas como MAE, RMSE e R², priorizando interpretabilidade e estabilidade dos coeficientes.

Obtenção de um R2 de 0.61 um possivel underfiting por se tratar de um modelo linear.
